rust/src/operations/merge.rs

```diff
@@ -0,0 +1,1614 @@
+
+    let current_metadata = snapshot
+        .current_metadata()
+        .ok_or(DeltaTableError::NoMetadata)?;
+
+    let predicate = match predicate {
+        Expression::DataFusion(expr) => expr,
+        Expression::String(s) => snapshot.parse_predicate_expression(s)?,
+    };
+
+    let schema = snapshot.input_schema()?;
+
+    // TODO: Given the join predicate, remove any expression that involve the
+    // source table and keep expressions that only involve the target table.
+    // This would allow us to perform statistics/partition pruning E.G
+    // Expression source.id = id and to_dt = '9999-12-31' -Becomes-> to_dt =
+    // '9999-12-31'
+    //
+    // If the user specified any not_source_match operations then those
+    // predicates also need to be considered when pruning
+
+    let target = parquet_scan_from_actions(
+        snapshot,
+        object_store.clone(),
+        snapshot.files(),
+        &schema,
+        None,
+        &state,
+        None,
+        None,
+    )
+    .await?;
+
+    let source = source.create_physical_plan().await?;
+
+    let source_count = Arc::new(MetricObserverExec::new(source, |batch, metrics| {
+        MetricBuilder::new(metrics)
+            .global_counter(SOURCE_COUNT_METRIC)
+            .add(batch.num_rows());
+    }));
+
+    let mut expressions: Vec<(Arc<dyn PhysicalExpr>, String)> = Vec::new();
+    let source_schema = source_count.schema();
+
+    let source_prefix = source_alias
+        .map(|mut s| {
+            s.push('.');
+            s
+        })
+        .unwrap_or_default();
+    for (i, field) in source_schema.fields().into_iter().enumerate() {
+        expressions.push((
+            Arc::new(expressions::Column::new(field.name(), i)),
+            source_prefix.clone() + field.name(),
+        ));
+    }
+    expressions.push((
+        Arc::new(expressions::Literal::new(true.into())),
+        "__delta_rs_source".to_owned(),
+    ));
+    let source = Arc::new(ProjectionExec::try_new(expressions, source_count.clone())?);
+
+    let mut expressions: Vec<(Arc<dyn PhysicalExpr>, String)> = Vec::new();
+    let target_schema = target.schema();
+    for (i, field) in target_schema.fields().into_iter().enumerate() {
+        expressions.push((
+            Arc::new(expressions::Column::new(field.name(), i)),
+            field.name().to_owned(),
+        ));
+    }
+    expressions.push((
+        Arc::new(expressions::Literal::new(true.into())),
+        "__delta_rs_target".to_owned(),
+    ));
+    let target = Arc::new(ProjectionExec::try_new(expressions, target.clone())?);
+
+    // TODO: Currently a NestedLoopJoin is used but we should target to support SortMergeJoin
+    // This would require rewriting the join predicate to only contain equality between left and right columns and pushing some filters down
+    // Ideally it would be nice if the optimizer / planner can pick the best join so maybe explore rewriting the entire operation using logical plans.
```

Comment:
Agreed. It would be nice if we could write this instead with the DataFrame API or as a logical plan, rather than having to write it as a physical one.